lib/Sema/TypeCheckType.cpp

```diff
@@ -1294,6 +1294,12 @@ static Type diagnoseUnknownType(TypeResolution resolution,
 
   // Unqualified lookup case.
   if (parentType.isNull()) {
+    if (resolution.getOptions().hasBase(TypeResolverContext::ExplicitCastExpr)) {
+      diags.diagnose(repr->getNameLoc(), diag::cannot_find_type_in_cast_expression,
+                     repr->getNameRef());
```

Comment:
I think we want to keep `cannot find type '...' in scope` in place if there was nothing found and only use new diagnostic for cases when we did find something which wasn't a type i.e. a variable use on the right-hand side of the operator.

Reply:
Presenting the new error message for type-casting expressions only - otherwise will fall to default `cannot find type "...' in scope`.

Reply:
ah ok I get it now - but I think this code is already presenting the new message only for cases where we find something that's not a type.

If I do something like:
`_ = a is` or `_ = a is struct` I get `error: expected type after 'is'`.
Only when I add something that's not a type I get the `type-casting operator expects...` message.

Reply:
> `_ = a is` or `_ = a is struct` I get `error: expected type after 'is'`

This is a parsing error. The error we want to improve on is emitted during semantic analysis, where the right-hand side of a syntactically valid type-casting expression is resolved to an existing type.

Consider the following test case:
```swift
func test(a: Any) {
    let _ = a is b
}
```
Here, the compiler will search for a declaration named `b` and will find neither an expected type declaration, nor anything else, so the expected error is `cannot find type 'b' in scope`.